examples/model_checking.py

```diff
@@ -35,115 +36,205 @@ def __init__(self, variables, init, trans):
-    return And(get_unrolling(system, k),
-               get_k_hypothesis(system, prop, k),
-               get_simple_path(system, k),
-               Not(prop_k))
-
-def check_property(system, prop):
-    """Interleaves BMC and K-Ind to verify the property."""
-    print("Checking property %s..." % prop)
-    for b in xrange(100):
-        f = get_bmc(system, prop, b)
-        print("   [BMC]    Checking bound %d..." % (b+1))
-        if is_sat(f):
-            print("--> Bug found at step %d" % (b+1))
-            return
-
-        f = get_k_induction(system, prop, b)
-        print("   [K-IND]  Checking bound %d..." % (b+1))
-        if is_unsat(f):
-            print("--> The system is safe!")
-            return
-
-
-def main():
-    # Example Transition System (SMV-like syntax)
+class PDR(object):
+    def __init__(self, system):
+        self.system = system
+        self.R = [system.init]
+        self.solver = Solver()
+        
+    def check_property(self, prop):
+        """Property Directed Reachability approach without optimizations."""
+        print("Checking property %s..." % prop)
+
+        while True:
+            cube = self.get_bad_state(prop)
+            if cube is not None:
+                if self.recursive_block(cube):
+                    print("--> Bug found at step %d" % (len(self.R)))
+                    return
+            else:
+                if self.inductive(prop):
+                    print("--> The system is safe!")
+                    return
+                print("   [PDR]    Adding frame %d..." % (len(self.R)))
+                self.R.append(TRUE())
+
+    def get_bad_state(self, prop):
+        """Extracts a reachable state that intersects the negation 
+        of the property and the last current frame"""
+        return self.solve(And(self.R[-1], Not(prop)))
+
+    def solve(self, formula):
+        """Provides a satisfiable assignment to the state variables that are consistent with the input formula"""
+        self.solver.reset_assertions()
+        self.solver.add_assertion(formula)
+        if self.solver.solve():
+            return And([EqualsOrIff(v, self.solver.get_value(v)) for v in self.system.variables])
+        return None
+
+    def recursive_block(self, cube):
+        """Blocks the cube at each frame, if possible."""
+        for i in range(len(self.R)-1, 0, -1):
+            cubeprime = cube.substitute(dict([(v, next_var(v)) for v in self.system.variables]))
+            cubepre = self.solve(And(self.R[i-1], self.system.trans, Not(cube), cubeprime))
+            if cubepre is None:
+                for j in range(1, i+1):
+                    self.R[j] = And(self.R[j], Not(cube))
+                return False
+            cube = cubepre
+        return True
+    
+    def inductive(self, prop):
+        """Checks if last two frames are equivalent """
+        if len(self.R) > 1 and \
+           self.solve(Not(EqualsOrIff(self.R[-1], self.R[-2]))) is None:
```

Comment:
You could use: ```self.solver.is_valid```.
self.R[-1] and self.R[-2] are boolean expressions. You can use Iff directly.